[libpas] Implement primary support for MTE

Achierius opened the pull request:
#### d3acbe4f96242ece76cca678aef67dbe30a53fc3
<pre>
[libpas] Implement primary support for MTE
<a href="https://bugs.webkit.org/show_bug.cgi?id=298803">https://bugs.webkit.org/show_bug.cgi?id=298803</a>
<a href="https://rdar.apple.com/160499929">rdar://160499929</a>

Reviewed by Mark Lam and Daniel Liu.

As announced on September 9th, the SoCs used in the next generation of
iPhones will include support for ARM&apos;S Memory Tagging Extension
functionality. As part of Apple&apos;s MIE (Memory Integrity Enforcement)
feature, libpas should thus implement support for MTE and related
memory-safety functionality to ensure that WebKit is up to par with the
new memory safety standards set by the rest of the system.

In particular, this patch ensures that when possible we allocate memory
with backing MTE tag pages and tag allocations made within them prior to
returning allocation memory to the caller. Not all memory can be tagged
this way: in particular, objects &gt;= 32K and objects which may be
referenced via compact pointers cannot be MTE tagged. There are other
exceptions as well, depending on process/object-type/platform.
It also implements a variety of hardening strategies to further
strengthen the feature and prevent certain well-known kinds of attacks.

Canonical link: <a href="https://commits.webkit.org/300469@main">https://commits.webkit.org/300469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75ad9aa5c8646b5d6fc709c8a751258c4a73c0f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122687 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129306 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/795b2a92-74de-4a32-8aff-a48ddcfcb87f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50992 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/49ca2bd2-0060-41d5-8807-b1d6ca82f678) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34375 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109834 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/73887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ed54229a-22eb-4f29-a434-4368fc2b2679) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/122044 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33362 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27990 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72793 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/114841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104074 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132036 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121216 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49632 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/101774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50009 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106048 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25807 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47011 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49489 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55242 "Built successfully") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151372 "Hash 75ad9aa5 for PR 50687 does not build (failure)") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/48956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/151372 "Hash 75ad9aa5 for PR 50687 does not build (failure)") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50639 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/84b900bbdebaaa570f4e9cfabb1c670013392f73. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/58b8f79e576cf9eb3e8e4ce7b5a28ff61c35f599. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dc0cbca38773c7b1e99aaf1157e0588eec4959cb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5a80a5f61980c55a5a2a9a04bc54f26314f7584d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7bb65554fe070637e898f2cd0670b7435a94b1ff. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/814f2af48eaa5a1d22fb013926a214d48e5cd7fb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a446557f20674fb8fb897a8de54fc647b31a7837. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/046ed9e574813085504643a3e97748e02586e1cd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5df38c80b03a0dc0e11e86305e6043989d3cb635. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a3cd791042a55bb2580a00e5ae7dcc9264dd1a35. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c91c52e2d6902833b279f7adb2037d286ffff374. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/eed8dba57d7adede9d7e6fbde1a09ff21514b002. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e99398a2ed4c78b2f0d1155b63bfcf0b74966075. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5a03d74881240e851555bf18b91854d7a2414321. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/be92d3debc6a65f84b4fb84bea85d66f2088fa77. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1637326d9e7fa1ab05b5e8e9f0747a16e7088cfb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4cda5ea53e65b812809a697f33aea54c5c92a341. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c93ce865582006ca29cc56371386fb7d3af7c0fe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7a0abc946d34804825c301cc4f978004f6e964eb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50687

Achierius committed:
[libpas] Implement primary support for MTE
https://bugs.webkit.org/show_bug.cgi?id=298803
rdar://160499929

Reviewed by Mark Lam and Daniel Liu.

As announced on September 9th, the SoCs used in the next generation of
iPhones will include support for ARM'S Memory Tagging Extension
functionality. As part of Apple's MIE (Memory Integrity Enforcement)
feature, libpas should thus implement support for MTE and related
memory-safety functionality to ensure that WebKit is up to par with the
new memory safety standards set by the rest of the system.

In particular, this patch ensures that when possible we allocate memory
with backing MTE tag pages and tag allocations made within them prior to
returning allocation memory to the caller. Not all memory can be tagged
this way: in particular, objects >= 32K and objects which may be
referenced via compact pointers cannot be MTE tagged. There are other
exceptions as well, depending on process/object-type/platform.
It also implements a variety of hardening strategies to further
strengthen the feature and prevent certain well-known kinds of attacks.

Canonical link: https://commits.webkit.org/300469@main
